src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs

```diff
@@ -0,0 +1,365 @@
+        }
+
+        private void ReportDiagnostic(Diagnostic d)
+        {
+            Flag();
+            _context.ReportDiagnostic(d);
+        }
+
+        ///<summary>Array types can only be one dimensional and not System.Array, 
+        ///and there are some types not usable in the Windows Runtime, like KeyValuePair</summary> 
+        ///<param name="typeSymbol">The type to check</param><param name="loc">where the type is</param>
+        ///<param name="memberId">The method or property with this type in its signature</param>
+        /// <param name="typeId">the type this member (method/prop) lives in</param>
+        private void ReportIfInvalidType(ITypeSymbol typeSymbol, Location loc, SyntaxToken memberId, SyntaxToken typeId) 
+        { 
+
+            if (typeSymbol.TypeKind == TypeKind.Array) 
+            { 
+                IsInvalidArrayType((IArrayTypeSymbol)typeSymbol, loc, memberId, typeId);
+                return;
+            }
+
+            foreach (var s in InvalidTypes)
+            { 
+                var invalidTypeSym = _context.Compilation.GetTypeByMetadataName(s);
+                if (SymbolEqualityComparer.Default.Equals(typeSymbol.OriginalDefinition, invalidTypeSym))
+                {
+                    Report(WinRTRules.UnsupportedTypeRule, loc, memberId, s, SuggestType(s));
+                    return;
+                }
+            }
+
+            // helper function, we can't get a type symbol for some of the invalid types 
+            bool SameType(ITypeSymbol sym, string s)
+            {
+                string baseStr = sym.ContainingNamespace.IsGlobalNamespace ? "" : sym.ContainingSymbol + ".";
+                var x = baseStr + sym.MetadataName;
+                return x == s;
+            }
+
+            // GetTypeByMetadataName fails on "System.Linq.Enumerable`1" & "System.Collections.ObjectModel.ReadOnlyDictionary`2"
+            // Would be fixed by issue #678 on the dotnet/roslyn-sdk repo
+            foreach (var invalidType in WIPInvalidTypes) 
+            {
+                if (SameType(typeSymbol, invalidType))
+                { 
+                    Report(WinRTRules.UnsupportedTypeRule, loc, memberId, invalidType, SuggestType(invalidType));
+                    return;
+                }
+            }
+        }
+ 
+        /// <summary>
+        /// See if this class/interfaces inherits the given type
+        /// </summary>
+        /// <param name="typeSymbol">type that might inherit</param>
+        /// <param name="typeToCheck">Inherited interface or class</param>
+        private bool ImplementsInterface(INamedTypeSymbol typeSymbol, string typeToCheck)
+        {
+            if (typeSymbol == null)
+            {
+                return false;
+            }
+
+            if (typeSymbol.BaseType != null)
+            {
+                return typeSymbol.BaseType.MetadataName == typeToCheck || typeSymbol.BaseType.ToString() == typeToCheck;

```

Comment:
A type can have a base type and also implement other interfaces.  Should we check interfaces if this is false?

Reply:
I had moved the interface loop up to before we check the base type. Not sure what you mean by check interfaces if this is false